crypto/rand/rand_unix.c

```diff
@@ -273,6 +283,129 @@ int syscall_random(void *buf, size_t buflen)
     return -1;
 }
 
+#if !defined(OPENSSL_RAND_SEED_NONE) && defined(OPENSSL_RAND_SEED_DEVRANDOM)
+static const char *random_device_paths[] = { DEVRANDOM };
+static struct random_device {
+    int fd;
+    dev_t dev;
+    ino_t ino;
+    mode_t mode;
+    dev_t rdev;
+} random_devices[OSSL_NELEM(random_device_paths)];
+static int keep_random_devices_open = 1;
+
+/*
+ * Verify that the file descriptor associated with the random source is
+ * still valid. The rationale for doing this is the fact that it is not
+ * uncommon for daemons to close all open file handles when daemonizing.
+ * So the handle might have been closed or even reused for opening
+ * another file.
+ */
+static int check_random_device(struct random_device * rd)
+{
+    struct stat st;
+
+    return rd->fd != -1
+           && fstat(rd->fd, &st) != -1
+           && rd->dev == st.st_dev
+           && rd->ino == st.st_ino
+           && rd->mode == st.st_mode
```

Comment:
One can wonder if it would be appropriate to mask off access bits. I mean should change of access permissions actually disqualify the descriptor?

Reply:
I think that you misunderstood the question. It's not whether or not there is reason for somebody to change access permissions on a device, or how any specific change would be justified to system owner. The permission can be changed, just use chmod, and the fact that they can change asks for question if changed permissions should be reason for disqualifying corresponding descriptor *here*.

Reply:
Since file access permissions cannot be withdrawn retroactively from an already open file handle, the logical consequence would be to ignore (mask) the permission bits.

Reply:
Updated.